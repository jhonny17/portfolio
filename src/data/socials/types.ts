export type Social = {
  id: number;
  name: string;
  url: string;
  icon: string;
};
