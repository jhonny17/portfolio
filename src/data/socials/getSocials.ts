import { default as socialsJson } from './socials.json';
import { Social } from './types';

export const getSocials = (): Social[] => {
  return socialsJson.socials as Social[];
};
