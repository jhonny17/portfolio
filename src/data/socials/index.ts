export * from './types';
export * from './getSocials';
