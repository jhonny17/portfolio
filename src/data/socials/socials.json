{
  "socials": [
    {
      "id": 1,
      "name": "LinkedIn",
      "url": "https://www.linkedin.com/in/jhonny17/",
      "icon": "linkedin"
    },
    {
      "id": 2,
      "name": "GitHub",
      "url": "https://github.com/jhonny17",
      "icon": "github"
    },
    {
      "id": 3,
      "name": "Discord",
      "url": "https://discordapp.com/users/jhonnyx17",
      "icon": "discord"
    }
  ]
}
