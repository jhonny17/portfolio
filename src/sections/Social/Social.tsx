import {
  IconBrandGithub,
  IconBrandDiscord,
  IconBrandLinkedin,
  IconProps,
  Icon,
} from '@tabler/icons-react';
import cx from 'classnames';

import { getSocials } from '@/data/socials';
import { ForwardRefExoticComponent, RefAttributes } from 'react';

const ICONS = {
  github: IconBrandGithub,
  discord: IconBrandDiscord,
  linkedin: IconBrandLinkedin,
} as Record<string, ForwardRefExoticComponent<IconProps & RefAttributes<Icon>>>;

export const Social = () => {
  const socials = getSocials();

  return (
    <section className={cx('w-full', 'bg-emerald-950')}>
      <h2 className="sr-only">Socials</h2>
      <div className={cx('flex', 'justify-center')}>
        {socials.map(({ id, icon, name, url }) => {
          const Icon = ICONS[icon];

          return (
            <a
              key={id}
              href={url}
              target="_blank"
              title={name}
              rel="noopener noreferrer"
              className={cx(
                'flex',
                'items-center',
                'justify-center',
                'w-12',
                'h-12',
                'hover:bg-neutral-900',
                'transition-colors',
                'duration-300',
                'ease-in-out',
              )}
            >
              <span className="sr-only">{name}</span>
              <Icon />
            </a>
          );
        })}
      </div>
    </section>
  );
};
