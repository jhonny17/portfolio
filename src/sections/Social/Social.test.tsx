import { expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import { Social } from './Social';
import { getSocials } from '@/data/socials';

type SocialsMap = {
  [key: string]: ReturnType<typeof getSocials>[number];
};

const socials: SocialsMap = getSocials().reduce((acc, social) => {
  acc[social.icon] = social;
  return acc;
}, {} as SocialsMap);

it('matches the snapshot', () => {
  const { asFragment } = render(<Social />);
  expect(asFragment()).toMatchSnapshot();
});

it('renders the social section', () => {
  render(<Social />);
  expect(
    screen.getByRole('heading', { level: 2, name: 'Socials' }),
  ).toBeInTheDocument();
});

it('renders the github icon', () => {
  render(<Social />);
  const social = socials['github'];
  expect(social).toBeDefined();

  const link = screen.getByRole('link', {
    name: social.name,
  }) as HTMLAnchorElement;

  expect(link).toBeInTheDocument();
  expect(link.href).toBe(social.url);
  expect(link.target).toBe('_blank');
  expect(link.rel).toBe('noopener noreferrer');
  expect(link.querySelector('svg')).toMatchSnapshot();
});

it('renders the discord icon', () => {
  render(<Social />);
  const social = socials['discord'];
  expect(social).toBeDefined();

  const link = screen.getByRole('link', {
    name: social.name,
  }) as HTMLAnchorElement;

  expect(link).toBeInTheDocument();
  expect(link.href).toBe(social.url);
  expect(link.target).toBe('_blank');
  expect(link.rel).toBe('noopener noreferrer');
  expect(link.querySelector('svg')).toMatchSnapshot();
});

it('renders the linkedin icon', () => {
  render(<Social />);
  const social = socials['linkedin'];
  expect(social).toBeDefined();

  const link = screen.getByRole('link', {
    name: social.name,
  }) as HTMLAnchorElement;

  expect(link).toBeInTheDocument();
  expect(link.href).toBe(social.url);
  expect(link.target).toBe('_blank');
  expect(link.rel).toBe('noopener noreferrer');
  expect(link.querySelector('svg')).toMatchSnapshot();
});
