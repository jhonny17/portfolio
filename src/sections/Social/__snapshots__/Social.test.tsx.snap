// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="w-full bg-emerald-950"
  >
    <h2
      class="sr-only"
    >
      Socials
    </h2>
    <div
      class="flex justify-center"
    >
      <a
        class="flex items-center justify-center w-12 h-12 hover:bg-neutral-900 transition-colors duration-300 ease-in-out"
        href="https://www.linkedin.com/in/jhonny17/"
        rel="noopener noreferrer"
        target="_blank"
        title="LinkedIn"
      >
        <span
          class="sr-only"
        >
          LinkedIn
        </span>
        <svg
          class="tabler-icon tabler-icon-brand-linkedin "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11v5"
          />
          <path
            d="M8 8v.01"
          />
          <path
            d="M12 16v-5"
          />
          <path
            d="M16 16v-3a2 2 0 1 0 -4 0"
          />
          <path
            d="M3 7a4 4 0 0 1 4 -4h10a4 4 0 0 1 4 4v10a4 4 0 0 1 -4 4h-10a4 4 0 0 1 -4 -4z"
          />
        </svg>
      </a>
      <a
        class="flex items-center justify-center w-12 h-12 hover:bg-neutral-900 transition-colors duration-300 ease-in-out"
        href="https://github.com/jhonny17"
        rel="noopener noreferrer"
        target="_blank"
        title="GitHub"
      >
        <span
          class="sr-only"
        >
          GitHub
        </span>
        <svg
          class="tabler-icon tabler-icon-brand-github "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 19c-4.3 1.4 -4.3 -2.5 -6 -3m12 5v-3.5c0 -1 .1 -1.4 -.5 -2c2.8 -.3 5.5 -1.4 5.5 -6a4.6 4.6 0 0 0 -1.3 -3.2a4.2 4.2 0 0 0 -.1 -3.2s-1.1 -.3 -3.5 1.3a12.3 12.3 0 0 0 -6.2 0c-2.4 -1.6 -3.5 -1.3 -3.5 -1.3a4.2 4.2 0 0 0 -.1 3.2a4.6 4.6 0 0 0 -1.3 3.2c0 4.6 2.7 5.7 5.5 6c-.6 .6 -.6 1.2 -.5 2v3.5"
          />
        </svg>
      </a>
      <a
        class="flex items-center justify-center w-12 h-12 hover:bg-neutral-900 transition-colors duration-300 ease-in-out"
        href="https://discordapp.com/users/jhonnyx17"
        rel="noopener noreferrer"
        target="_blank"
        title="Discord"
      >
        <span
          class="sr-only"
        >
          Discord
        </span>
        <svg
          class="tabler-icon tabler-icon-brand-discord "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 12a1 1 0 1 0 2 0a1 1 0 0 0 -2 0"
          />
          <path
            d="M14 12a1 1 0 1 0 2 0a1 1 0 0 0 -2 0"
          />
          <path
            d="M15.5 17c0 1 1.5 3 2 3c1.5 0 2.833 -1.667 3.5 -3c.667 -1.667 .5 -5.833 -1.5 -11.5c-1.457 -1.015 -3 -1.34 -4.5 -1.5l-.972 1.923a11.913 11.913 0 0 0 -4.053 0l-.975 -1.923c-1.5 .16 -3.043 .485 -4.5 1.5c-2 5.667 -2.167 9.833 -1.5 11.5c.667 1.333 2 3 3.5 3c.5 0 2 -2 2 -3"
          />
          <path
            d="M7 16.5c3.5 1 6.5 1 10 0"
          />
        </svg>
      </a>
    </div>
  </section>
</DocumentFragment>
`;

exports[`renders the discord icon 1`] = `
<svg
  class="tabler-icon tabler-icon-brand-discord "
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M8 12a1 1 0 1 0 2 0a1 1 0 0 0 -2 0"
  />
  <path
    d="M14 12a1 1 0 1 0 2 0a1 1 0 0 0 -2 0"
  />
  <path
    d="M15.5 17c0 1 1.5 3 2 3c1.5 0 2.833 -1.667 3.5 -3c.667 -1.667 .5 -5.833 -1.5 -11.5c-1.457 -1.015 -3 -1.34 -4.5 -1.5l-.972 1.923a11.913 11.913 0 0 0 -4.053 0l-.975 -1.923c-1.5 .16 -3.043 .485 -4.5 1.5c-2 5.667 -2.167 9.833 -1.5 11.5c.667 1.333 2 3 3.5 3c.5 0 2 -2 2 -3"
  />
  <path
    d="M7 16.5c3.5 1 6.5 1 10 0"
  />
</svg>
`;

exports[`renders the github icon 1`] = `
<svg
  class="tabler-icon tabler-icon-brand-github "
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M9 19c-4.3 1.4 -4.3 -2.5 -6 -3m12 5v-3.5c0 -1 .1 -1.4 -.5 -2c2.8 -.3 5.5 -1.4 5.5 -6a4.6 4.6 0 0 0 -1.3 -3.2a4.2 4.2 0 0 0 -.1 -3.2s-1.1 -.3 -3.5 1.3a12.3 12.3 0 0 0 -6.2 0c-2.4 -1.6 -3.5 -1.3 -3.5 -1.3a4.2 4.2 0 0 0 -.1 3.2a4.6 4.6 0 0 0 -1.3 3.2c0 4.6 2.7 5.7 5.5 6c-.6 .6 -.6 1.2 -.5 2v3.5"
  />
</svg>
`;

exports[`renders the linkedin icon 1`] = `
<svg
  class="tabler-icon tabler-icon-brand-linkedin "
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M8 11v5"
  />
  <path
    d="M8 8v.01"
  />
  <path
    d="M12 16v-5"
  />
  <path
    d="M16 16v-3a2 2 0 1 0 -4 0"
  />
  <path
    d="M3 7a4 4 0 0 1 4 -4h10a4 4 0 0 1 4 4v10a4 4 0 0 1 -4 4h-10a4 4 0 0 1 -4 -4z"
  />
</svg>
`;
